sockeye/decoder.py

```diff
@@ -14,85 +14,123 @@
 """
 Decoders for sequence-to-sequence models.
 """
+import logging
 from typing import Callable, List, NamedTuple, Tuple
 from typing import Optional
-import logging
 
 import mxnet as mx
 
 from sockeye.config import Config
-from sockeye.layers import LayerNormalization
 from sockeye.utils import check_condition
 from . import attention as attentions
 from . import constants as C
 from . import encoder
+from . import layers
 from . import lexicon as lexicons
 from . import rnn
-
+from . import transformer
 
 logger = logging.getLogger(__name__)
 
 
-class RecurrentDecoderConfig(Config):
-    """
-    Recurrent decoder configuration.
+def get_decoder(config: Config,
+                lexicon: Optional[lexicons.Lexicon] = None,
+                embed_weight: Optional[mx.sym.Symbol] = None) -> 'Decoder':
+    if isinstance(config, RecurrentDecoderConfig):
+        return RecurrentDecoder(config=config, lexicon=lexicon, embed_weight=embed_weight, prefix=C.DECODER_PREFIX)
+    elif isinstance(config, transformer.TransformerConfig):
+        return TransformerDecoder(config=config, embed_weight=embed_weight, prefix=C.DECODER_PREFIX)
+    else:
+        raise ValueError("Unsupported decoder configuration")
 
-    :param vocab_size: Target vocabulary size.
-    :param num_embed: Target word embedding size.
-    :param rnn_config: RNN configuration.
-    :param dropout: Dropout probability for decoder RNN.
-    :param weight_tying: Whether to share embedding and prediction parameter matrices.
-    :param context_gating: Whether to use context gating.
-    :param layer_normalization: Apply layer normalization.
+
+class Decoder:
+    """
+    Generic decoder interface.
+    A decoder needs to provide methods to provide training (decode) and inference symbols (predict).
```

Comment:
I will change the wording to make this a bit clearer, fully agreed that this is not specific to inference/search